File: auth-authorization-server-api/src/main/java/com/cartobucket/auth/authorization/server/routes/AuthorizationServer.java
/*
 * Copyright 2023 Bryce Groff (Cartobucket LLC)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.cartobucket.auth.authorization.server.routes;

import com.cartobucket.auth.authorization.server.routes.mappers.AccessTokenResponseMapper;
import com.cartobucket.auth.authorization.server.routes.mappers.JwksMapper;
import com.cartobucket.auth.data.domain.ClientCode;
import com.cartobucket.auth.data.domain.Page;
import com.cartobucket.auth.data.domain.Scope;
import com.cartobucket.auth.data.services.ApplicationService;
import com.cartobucket.auth.data.services.ScopeService;
import com.cartobucket.auth.generated.AuthorizationServerApi;
import com.cartobucket.auth.model.generated.AccessTokenRequest;
import com.cartobucket.auth.data.services.AuthorizationServerService;
import com.cartobucket.auth.data.services.ClientService;
import com.cartobucket.auth.data.services.TemplateService;
import com.cartobucket.auth.data.services.UserService;
import jakarta.ws.rs.BadRequestException;
import jakarta.ws.rs.Consumes;
import jakarta.ws.rs.FormParam;
import jakarta.ws.rs.GET;
import jakarta.ws.rs.POST;
import jakarta.ws.rs.Path;
import jakarta.ws.rs.core.Response;
import org.jboss.logging.Logger;

import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

public abstract class AuthorizationServer implements AuthorizationServerApi {
    private static final Logger LOG = Logger.getLogger(AuthorizationServer.class);

    final AuthorizationServerService authorizationServerService;
    final ClientService clientService;
    final UserService userService;
    final TemplateService templateService;
    final ApplicationService applicationService;
    final ScopeService scopeService;

    public AuthorizationServer(
            AuthorizationServerService authorizationServerService,
            ClientService clientService,
            UserService userService, TemplateService templateService,
            ApplicationService applicationService, ScopeService scopeService) {
        this.authorizationServerService = authorizationServerService;
        this.clientService = clientService;
        this.userService = userService;
        this.templateService = templateService;
        this.applicationService = applicationService;
        this.scopeService = scopeService;
    }

    @Override
    @Consumes({"*/*"})
    public Response createAuthorizationCode(
            UUID authorizationServerId,
            String clientId,
            String responseType,
            String codeChallenge,
            String codeChallengeMethod,
            String redirectUri,
            String scope,
            String state,
            String nonce,
            String username,
            String password) {
        // TODO: Check the CSRF token

        final var user = userService.getUser(username).getLeft();
        if (user == null) {
            LOG.error("User not found: " + username);
            return renderLoginScreen(authorizationServerId);
        }
        if (!user.getAuthorizationServerId().equals(authorizationServerId)) {
            LOG.error("AuthorizationServer does not equal the user AuthorizationServer: " + authorizationServerId);
            return renderLoginScreen(authorizationServerId);
        }
        if (!userService.validatePassword(user.getId(), password)) {
            LOG.error("Password was not correct: " + password);
            return renderLoginScreen(authorizationServerId);
        }
        final var client = clientService.getClient(clientId);
        if (client == null) {
            LOG.error("Client not found: " + clientId);
            return renderLoginScreen(authorizationServerId);
        }
        if (!client.getAuthorizationServerId().equals(authorizationServerId)) {
            LOG.error("Client not found in AuthorizationServer: " + authorizationServerId);
            return renderLoginScreen(authorizationServerId);
        }
        if (client.getRedirectUris().contains(redirectUri)) {
            LOG.error("Redirects not found: " + redirectUri);
            return renderLoginScreen(authorizationServerId);
        }

        // TODO: FIx this
//        final var scopes = ScopeService.scopeStringToScopeList(scope);
//        if (client.getScopes().containsAll(scopes)) {
//            LOG.error("Scopes not found: " + scopes + " and " + client.getScopes().toString());
//            return renderLoginScreen(authorizationServerId);
//        }

        var clientCode = new ClientCode();
        clientCode.setClientId(clientId);
        clientCode.setRedirectUri(redirectUri);
        clientCode.setAuthorizationServerId(authorizationServerId);
        clientCode.setNonce(nonce);
        clientCode.setState(state);
        clientCode.setCodeChallenge(codeChallenge);
        clientCode.setCodeChallengeMethod(codeChallengeMethod);
        clientCode.setScopes(ScopeService.scopeStringToScopeList(scope));
        clientCode.setUserId(user.getId());

        LOG.error("We are about to call into the client service now");
        final var code = clientService.createClientCode(
                authorizationServerId,
                clientCode
        );

        return Response.status(302).location(
                URI.create(
                        redirectUri +
                                "?code=" +
                                URLEncoder.encode(code.getCode(), StandardCharsets.UTF_8) +
                                "&state=" + URLEncoder.encode(code.getState(), StandardCharsets.UTF_8) +
                                "&nonce=" + URLEncoder.encode(code.getNonce(), StandardCharsets.UTF_8) +
                                "&scope" + URLEncoder.encode("openid", StandardCharsets.UTF_8)
//                                        ScopeService.scopeListToScopeString(
//                                                code
//                                                        .getScopes()
//                                                        .stream()
//                                                        .map(Scope::getName)
//                                                        .toList()
//                                        ), StandardCharsets.UTF_8)
                )
        ).build();
    }

    @Override
    public Response getAuthorizationServerJwks(UUID authorizationServerId) {
        return Response
                .ok()
                .entity(
                        JwksMapper.toJwksResponse(
                                authorizationServerService.getJwksForAuthorizationServer(authorizationServerId)
                        )
                )
                .build();
    }

    @Path(".well-known/openid-configuration")
    @GET
    @Consumes({"*/*"})
    public Response getOpenIdConfiguration(UUID authorizationServerId) {
        return getOpenIdConnectionWellKnown(authorizationServerId);
    }

    @Override
    public Response getOpenIdConnectionWellKnown(UUID authorizationServerId) {
        final var authorizationServer = authorizationServerService.getAuthorizationServer(authorizationServerId);
        var wellKnown = new com.cartobucket.auth.model.generated.WellKnown();
        wellKnown.setIssuer(authorizationServer.getServerUrl() + "/" + authorizationServer.getId() + "/");
        wellKnown.setAuthorizationEndpoint(
                authorizationServer.getServerUrl().toString() + "/"  + authorizationServer.getId() + "/authorization/");
        wellKnown.setTokenEndpoint(
                authorizationServer.getServerUrl().toString() + "/"  + authorizationServer.getId() + "/token/");
        wellKnown.setJwksUri(
                authorizationServer.getServerUrl().toString() + "/"  + authorizationServer.getId() + "/jwks/");
        wellKnown.setRevocationEndpoint(
                authorizationServer.getServerUrl().toString() + "/"  + authorizationServer.getId() + "/revocation/");
        wellKnown.setUserinfoEndpoint(
                authorizationServer.getServerUrl().toString() + "/"  + authorizationServer.getId() + "/userinfo/");
        wellKnown.setTokenEndpointAuthMethodsSupported(
                List.of(
                        com.cartobucket.auth.model.generated.WellKnown.TokenEndpointAuthMethodsSupportedEnum.POST
                )
        );
        wellKnown.setIdTokenSigningAlgValuesSupported(
                List.of(
                        com.cartobucket.auth.model.generated.WellKnown.IdTokenSigningAlgValuesSupportedEnum.RS256
                )
        );
        wellKnown.setResponseTypesSupported(
                Arrays.asList(
                        com.cartobucket.auth.model.generated.WellKnown.ResponseTypesSupportedEnum.CODE,
                        com.cartobucket.auth.model.generated.WellKnown.ResponseTypesSupportedEnum.CODE_ID_TOKEN,
                        com.cartobucket.auth.model.generated.WellKnown.ResponseTypesSupportedEnum.TOKEN
                )
        );
        wellKnown.setCodeChallengeMethodsSupported(List.of(com.cartobucket.auth.model.generated.WellKnown.CodeChallengeMethodsSupportedEnum.S256));
        wellKnown.setGrantTypesSupported(
                Arrays.asList(
                        AccessTokenRequest.GrantTypeEnum.CLIENT_CREDENTIALS.value(),
                        AccessTokenRequest.GrantTypeEnum.AUTHORIZATION_CODE.value()
                )
        );
        return Response.ok().entity(wellKnown).build();
    }

    @Override
    public Response getUserInfo(UUID authorizationServerId, String authorization) {
        final var jwtClaims = authorizationServerService.validateJwtForAuthorizationServer(
                authorizationServerId,
                authorization
        );
        return Response
                .ok()
                .entity(userService.getUser(UUID.fromString(String.valueOf(jwtClaims.get("sub")))).getRight().getProfile())
                .build();
    }

    @Override
    @Consumes({ "*/*" })
    public Response initiateAuthorization(UUID authorizationServerId, String clientId, String responseType, String codeChallenge, String codeChallengeMethod, String redirectUri, String scope, String state, String nonce) {
        return renderLoginScreen(authorizationServerId);
    }

    @POST
    @Path("/token/")
    @Consumes({"application/x-www-form-urlencoded"})
    public Response issueToken(
            UUID authorizationServerId,
            @FormParam("client_id") String clientId,
            @FormParam("client_secret") String clientSecret,
            @FormParam("grant_type") String grantType,
            @FormParam("code") String code,
            @FormParam("redirect_uri") String redirectUri,
            @FormParam("code_verifier") String codeVerifier,
            @FormParam("scope") String scope) {
        final var accessTokenRequest = new AccessTokenRequest();
        accessTokenRequest.setClientId(clientId);
        accessTokenRequest.setClientSecret(clientSecret);
        accessTokenRequest.setGrantType(AccessTokenRequest.GrantTypeEnum.fromValue(grantType));
        accessTokenRequest.setCode(code);
        accessTokenRequest.setRedirectUri(redirectUri);
        accessTokenRequest.setCodeVerifier(codeVerifier);
        accessTokenRequest.setScope(scope != null ? scope : "");
        return issueToken(authorizationServerId, accessTokenRequest);
    }

    @Override
    public Response issueToken(UUID authorizationServerId, AccessTokenRequest accessTokenRequest) {
        final var authorizationServer = authorizationServerService.getAuthorizationServer(authorizationServerId);

        // TODO: This stuff should be moved to a validator
        switch (accessTokenRequest.getGrantType()) {
            case CLIENT_CREDENTIALS -> {
                if (applicationService.isApplicationSecretValid(
                        authorizationServerId,
                        UUID.fromString(accessTokenRequest.getClientId()),
                        accessTokenRequest.getClientSecret())) {
                    throw new BadRequestException();
                }

                final var scopes = scopeService.filterScopesForAuthorizationServerId(
                        authorizationServerId,
                        accessTokenRequest.getScope()
                );

                return Response
                        .ok()
                        .entity(
                                authorizationServerService
                                        .generateAccessToken(
                                                authorizationServerId,
                                                UUID.fromString(accessTokenRequest.getClientId()),
                                                accessTokenRequest.getClientId(),
                                                // TODO: Fix this
                                                scopeService.getScopes(List.of(authorizationServerId), new Page(100, 0)),
                                                //scopes,
                                                authorizationServer.getClientCredentialsTokenExpiration(),
                                                null
                                        )
                        )
                        .build();
            }
            case AUTHORIZATION_CODE -> {
                final var clientCode = clientService.getClientCode(accessTokenRequest.getCode());
                if (clientCode == null) {
                    throw new BadRequestException();
                }

                if (!clientCode.getRedirectUri().equals(accessTokenRequest.getRedirectUri())) {
                    throw new BadRequestException();
                }

                return Response
                        .ok()
                        .entity(
                                AccessTokenResponseMapper.toAccessTokenResponse(
                                        authorizationServerService
                                                .generateAccessToken(
                                                        authorizationServerId,
                                                        clientCode.getUserId(),
                                                        String.valueOf(clientCode.getUserId()),
                                                        clientCode.getScopes(),
                                                        authorizationServer.getAuthorizationCodeTokenExpiration(),
                                                        clientCode.getNonce()
                                                )
                                )
                        )
                        .build();
            }
        }
        throw new BadRequestException();
    }

    protected abstract Response renderLoginScreen(UUID authorizationServerId);
}

File: auth-authorization-server-api/src/main/kotlin/com/cartobucket/auth/authorization/server/routes/AuthorizationServer.kt
/*
 * Copyright 2023 Bryce Groff (Cartobucket LLC)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions
 * of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.cartobucket.auth.authorization.server.routes

import com.cartobucket.auth.authorization.server.dto.AccessTokenRequest
import com.cartobucket.auth.authorization.server.dto.WellKnown
import com.cartobucket.auth.authorization.server.interfaces.AuthorizationServerApi
import com.cartobucket.auth.authorization.server.routes.mappers.AccessTokenResponseMapper.toAccessTokenResponse
import com.cartobucket.auth.authorization.server.routes.mappers.JwksMapper.toJwksResponse
import com.cartobucket.auth.data.domain.ClientCode
import com.cartobucket.auth.data.domain.Scope
import com.cartobucket.auth.data.services.*
import jakarta.ws.rs.*
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response
import org.jboss.logging.Logger
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import java.util.*

abstract class AuthorizationServer(
    protected val authorizationServerService: AuthorizationServerService,
    protected val clientService: ClientService,
    protected val userService: UserService,
    protected val templateService: TemplateService,
    protected val applicationService: ApplicationService,
    protected val scopeService: ScopeService
) : AuthorizationServerApi {

    companion object {
        private val LOG = Logger.getLogger(AuthorizationServer::class.java)
    }

    @Consumes("multipart/form-data", MediaType.APPLICATION_FORM_URLENCODED, MediaType.APPLICATION_JSON)
    override fun createAuthorizationCode(
        authorizationServerId: UUID,
        clientId: String?,
        responseType: String?,
        codeChallenge: String?,
        codeChallengeMethod: String?,
        redirectUri: String?,
        scope: String?,
        state: String?,
        nonce: String?,
        username: String?,
        password: String?
    ): Response {
        // If no username/password provided, show login screen
        if (username.isNullOrEmpty() || password.isNullOrEmpty()) {
            return renderLoginScreen(authorizationServerId)
        }

        val user = try {
            userService.getUser(username!!).left
        } catch (e: Exception) {
            LOG.error("User not found: $username")
            return renderLoginScreen(authorizationServerId, "Invalid username or password")
        }

        if (user == null) {
            LOG.error("User not found: $username")
            return renderLoginScreen(authorizationServerId, "Invalid username or password")
        }
        
        if (user.authorizationServerId != authorizationServerId) {
            LOG.error("AuthorizationServer does not equal the user AuthorizationServer: $authorizationServerId")
            return renderLoginScreen(authorizationServerId, "Invalid username or password")
        }
        
        if (!userService.validatePassword(user.id!!, password!!)) {
            LOG.error("Password was not correct")
            return renderLoginScreen(authorizationServerId, "Invalid username or password")
        }

        val client = try {
            clientService.getClient(clientId!!)
        } catch (e: com.cartobucket.auth.data.exceptions.notfound.ClientNotFound) {
            LOG.error("Client not found in POST: $clientId")
            return renderLoginScreen(authorizationServerId, "Invalid client_id")
        }

        if (client.authorizationServerId != authorizationServerId) {
            LOG.error("Client not authorized for server in POST: $clientId")
            return renderLoginScreen(authorizationServerId, "Invalid client_id")
        }

        val validScopes = scopeService.filterScopesForAuthorizationServerId(
            authorizationServerId,
            scope ?: ""
        )

        val clientCode = ClientCode().apply {
            this.clientId = clientId
            this.redirectUri = redirectUri
            this.authorizationServerId = authorizationServerId
            this.nonce = nonce
            this.state = state
            this.codeChallenge = codeChallenge
            this.codeChallengeMethod = codeChallengeMethod
            this.scopes = validScopes
            this.userId = user.id!!
        }

        LOG.error("We are about to call into the client service now")
        val code = clientService.createClientCode(authorizationServerId, clientCode)

        var redirectUrl = "${redirectUri}?code=${URLEncoder.encode(code.code, StandardCharsets.UTF_8)}"

        code.state?.let {
            redirectUrl += "&state=${URLEncoder.encode(it, StandardCharsets.UTF_8)}"
        }

        code.nonce?.let {
            redirectUrl += "&nonce=${URLEncoder.encode(it, StandardCharsets.UTF_8)}"
        }

        if (!code.scopes.isNullOrEmpty()) {
            val scopeString = ScopeService.scopeListToScopeString(
                code.scopes!!.map { it.name!! }
            )
            redirectUrl += "&scope=${URLEncoder.encode(scopeString, StandardCharsets.UTF_8)}"
        } else {
            redirectUrl += "&scope=${URLEncoder.encode("openid", StandardCharsets.UTF_8)}"
        }

        return Response.status(302)
            .location(URI.create(redirectUrl))
            .build()
    }

    override fun getAuthorizationServerJwks(authorizationServerId: UUID): Response {
        return Response
            .ok()
            .entity(
                authorizationServerService.getJwksForAuthorizationServer(authorizationServerId)
                    .toJwksResponse()
            )
            .build()
    }

    @Path(".well-known/openid-configuration")
    @GET
    fun getOpenIdConfiguration(authorizationServerId: UUID): Response =
        getOpenIdConnectionWellKnown(authorizationServerId)

    override fun getOpenIdConnectionWellKnown(authorizationServerId: UUID): Response {
        val authorizationServer = authorizationServerService.getAuthorizationServer(authorizationServerId)
        val wellKnown = WellKnown(
            issuer = "${authorizationServer.serverUrl}/${authorizationServer.id}/",
            authorizationEndpoint = "${authorizationServer.serverUrl}/${authorizationServer.id}/authorization/",
            tokenEndpoint = "${authorizationServer.serverUrl}/${authorizationServer.id}/token/",
            jwksUri = "${authorizationServer.serverUrl}/${authorizationServer.id}/jwks/",
            revocationEndpoint = "${authorizationServer.serverUrl}/${authorizationServer.id}/revocation/",
            userinfoEndpoint = "${authorizationServer.serverUrl}/${authorizationServer.id}/userinfo/",
            tokenEndpointAuthMethodsSupported = listOf(WellKnown.TokenEndpointAuthMethodsSupportedEnum.POST.value),
            idTokenSigningAlgValuesSupported = listOf(WellKnown.IdTokenSigningAlgValuesSupportedEnum.RS256.value),
            responseTypesSupported = listOf(
                WellKnown.ResponseTypesSupportedEnum.CODE.value,
                WellKnown.ResponseTypesSupportedEnum.CODE_ID_TOKEN.value,
                WellKnown.ResponseTypesSupportedEnum.TOKEN.value
            ),
            codeChallengeMethodsSupported = listOf(WellKnown.CodeChallengeMethodsSupportedEnum.S256.value),
            grantTypesSupported = listOf(
                AccessTokenRequest.GrantTypeEnum.CLIENT_CREDENTIALS.value,
                AccessTokenRequest.GrantTypeEnum.AUTHORIZATION_CODE.value,
                AccessTokenRequest.GrantTypeEnum.REFRESH_TOKEN.value
            )
        )
        return Response.ok().entity(wellKnown).build()
    }

    override fun getUserInfo(authorizationServerId: UUID, authorization: String?): Response {
        val jwtClaims = authorizationServerService.validateJwtForAuthorizationServer(
            authorizationServerId,
            authorization!!
        )
        return Response
            .ok()
            .entity(
                userService.getUser(UUID.fromString(jwtClaims["sub"].toString())).right?.profile
            )
            .build()
    }

    override fun getAuthorization(
        authorizationServerId: UUID,
        clientId: String?,
        responseType: String?,
        redirectUri: String?,
        scope: String?,
        state: String?,
        nonce: String?,
        codeChallenge: String?,
        codeChallengeMethod: String?
    ): Response = initiateAuthorization(
        authorizationServerId,
        clientId,
        responseType,
        codeChallenge,
        codeChallengeMethod,
        redirectUri,
        scope,
        state,
        nonce
    )

    private fun initiateAuthorization(
        authorizationServerId: UUID,
        clientId: String?,
        responseType: String?,
        codeChallenge: String?,
        codeChallengeMethod: String?,
        redirectUri: String?,
        scope: String?,
        state: String?,
        nonce: String?
    ): Response {
        // Validate client_id first
        if (clientId.isNullOrEmpty()) {
            return Response.status(Response.Status.BAD_REQUEST)
                .entity("""{"error":"invalid_request","error_description":"client_id is required"}""")
                .type(MediaType.APPLICATION_JSON)
                .build()
        }

        val client = try {
            clientService.getClient(clientId)
        } catch (e: com.cartobucket.auth.data.exceptions.notfound.ClientNotFound) {
            return Response.status(Response.Status.UNAUTHORIZED)
                .entity("""{"error":"invalid_client","error_description":"Client authentication failed"}""")
                .type(MediaType.APPLICATION_JSON)
                .build()
        }

        if (client.authorizationServerId != authorizationServerId) {
            return Response.status(Response.Status.UNAUTHORIZED)
                .entity("""{"error":"invalid_client","error_description":"Client not authorized for this server"}""")
                .type(MediaType.APPLICATION_JSON)
                .build()
        }

        // Validate redirect_uri
        if (redirectUri.isNullOrEmpty()) {
            return Response.status(Response.Status.BAD_REQUEST)
                .entity("""{"error":"invalid_request","error_description":"redirect_uri is required"}""")
                .type(MediaType.APPLICATION_JSON)
                .build()
        }

        val redirectUriAllowed = try {
            val requestedUri = URI.create(redirectUri)
            client.redirectUris?.any { allowedUri ->
                try {
                    normalizeUri(allowedUri) == normalizeUri(requestedUri)
                } catch (e: Exception) {
                    LOG.warn("Invalid URI in client redirect URIs: $allowedUri")
                    false
                }
            } ?: false
        } catch (e: Exception) {
            return Response.status(Response.Status.BAD_REQUEST)
                .entity("""{"error":"invalid_request","error_description":"Invalid redirect_uri format"}""")
                .type(MediaType.APPLICATION_JSON)
                .build()
        }

        if (!redirectUriAllowed) {
            return Response.status(Response.Status.BAD_REQUEST)
                .entity("""{"error":"invalid_request","error_description":"redirect_uri not registered for this client"}""")
                .type(MediaType.APPLICATION_JSON)
                .build()
        }

        // Validate scopes
        if (!scope.isNullOrEmpty()) {
            val requestedScopes = ScopeService.scopeStringToScopeList(scope)
            val validScopes = scopeService.filterScopesForAuthorizationServerId(
                authorizationServerId,
                scope
            )

            val validScopeNames = validScopes.map { it.name!! }
            val invalidScopeNames = requestedScopes
                .map { it.name!! }
                .filter { scopeName -> !validScopeNames.contains(scopeName) }

            if (invalidScopeNames.isNotEmpty()) {
                val errorUri = redirectUri +
                    (if (redirectUri.contains("?")) "&" else "?") +
                    "error=invalid_scope" +
                    "&error_description=" + URLEncoder.encode(
                        "Invalid scopes: ${invalidScopeNames.joinToString(", ")}",
                        StandardCharsets.UTF_8
                    ) +
                    (state?.let { "&state=$it" } ?: "")
                
                return Response.status(Response.Status.FOUND)
                    .location(URI.create(errorUri))
                    .build()
            }
        }

        return renderLoginScreen(authorizationServerId)
    }

    @POST
    @Path("/token/")
    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
    fun issueToken(
        authorizationServerId: UUID,
        @FormParam("client_id") clientId: String?,
        @FormParam("client_secret") clientSecret: String?,
        @FormParam("grant_type") grantType: String?,
        @FormParam("code") code: String?,
        @FormParam("redirect_uri") redirectUri: String?,
        @FormParam("code_verifier") codeVerifier: String?,
        @FormParam("scope") scope: String?,
        @FormParam("refresh_token") refreshToken: String?
    ): Response = postToken(
        authorizationServerId,
        grantType,
        clientId,
        clientSecret,
        code,
        redirectUri,
        codeVerifier,
        refreshToken,
        scope
    )

    override fun postTokenJson(authorizationServerId: UUID, request: AccessTokenRequest?): Response =
        postToken(
            authorizationServerId,
            request?.grantType?.value,
            request?.clientId,
            request?.clientSecret,
            request?.code,
            request?.redirectUri,
            request?.codeVerifier,
            request?.refreshToken,
            request?.scope
        )

    override fun postToken(
        authorizationServerId: UUID,
        grantType: String?,
        clientId: String?,
        clientSecret: String?,
        code: String?,
        redirectUri: String?,
        codeVerifier: String?,
        refreshToken: String?,
        scope: String?
    ): Response {
        val authorizationServer = authorizationServerService.getAuthorizationServer(authorizationServerId)
        val grantTypeEnum = AccessTokenRequest.GrantTypeEnum.fromValue(grantType!!)

        return when (grantTypeEnum) {
            AccessTokenRequest.GrantTypeEnum.CLIENT_CREDENTIALS -> {
                if (!applicationService.isApplicationSecretValid(
                        authorizationServerId,
                        UUID.fromString(clientId!!),
                        clientSecret!!
                    )) {
                    throw BadRequestException()
                }

                val applicationSecret = applicationService.getApplicationSecret(clientId!!)
                val scopes = scopeService.filterScopesForAuthorizationServerId(
                    authorizationServerId,
                    scope ?: ""
                ).filter { requestedScope ->
                    applicationSecret.scopes?.contains(requestedScope) == true
                }

                Response.ok()
                    .entity(
                        authorizationServerService.generateClientCredentialsAccessToken(
                            authorizationServerId,
                            applicationSecret.applicationId!!,
                            applicationSecret.applicationId.toString(),
                            scopes,
                            authorizationServer.clientCredentialsTokenExpiration!!
                        ).toAccessTokenResponse()
                    )
                    .build()
            }

            AccessTokenRequest.GrantTypeEnum.AUTHORIZATION_CODE -> {
                val clientCode = clientService.getClientCode(code!!)
                    ?: throw BadRequestException()

                if (clientCode.redirectUri != redirectUri) {
                    throw BadRequestException()
                }

                Response.ok()
                    .entity(
                        authorizationServerService.generateAuthorizationCodeFlowAccessToken(
                            authorizationServerId,
                            clientCode.userId!!,
                            clientCode.userId.toString(),
                            clientCode.clientId!!,
                            clientCode.scopes ?: emptyList(),
                            authorizationServer.authorizationCodeTokenExpiration!!,
                            clientCode.nonce
                        ).toAccessTokenResponse()
                    )
                    .build()
            }

            AccessTokenRequest.GrantTypeEnum.REFRESH_TOKEN -> {
                if (refreshToken.isNullOrEmpty()) {
                    throw BadRequestException("refresh_token is required")
                }

                if (clientId.isNullOrEmpty()) {
                    throw BadRequestException("client_id is required")
                }

                Response.ok()
                    .entity(
                        authorizationServerService.refreshAccessToken(
                            authorizationServerId,
                            refreshToken!!,
                            clientId!!
                        ).toAccessTokenResponse()
                    )
                    .build()
            }
        }
    }

    protected abstract fun renderLoginScreen(authorizationServerId: UUID): Response

    protected open fun renderLoginScreen(authorizationServerId: UUID, errorMessage: String): Response =
        renderLoginScreen(authorizationServerId)

    private fun normalizeUri(uri: URI): String {
        val scheme = uri.scheme?.lowercase() ?: ""
        val host = uri.host?.lowercase() ?: ""
        var port = uri.port
        val path = uri.path ?: ""
        val query = uri.query
        val fragment = uri.fragment

        // Remove default ports
        if ((port == 80 && scheme == "http") || (port == 443 && scheme == "https")) {
            port = -1
        }

        return buildString {
            append(scheme)
            if (scheme.isNotEmpty()) {
                append("://")
            }
            append(host)
            if (port != -1) {
                append(":").append(port)
            }
            append(path)
            query?.let { append("?").append(it) }
            fragment?.let { append("#").append(it) }
        }
    }
}